text/3503-frontmatter.md

```diff
@@ -0,0 +1,486 @@
+- Feature Name: `frontmatter`
+- Start Date: 2023-09-26
+- RFC PR: [rust-lang/rfcs#3503](https://github.com/rust-lang/rfcs/pull/3503)
+- Rust Issue: [rust-lang/cargo#12207](https://github.com/rust-lang/cargo/issues/12207)
+
+
+# Summary
+[summary]: #summary
+
+Add a frontmatter syntax to Rust as a way for [cargo to have manifests embedded in source code](https://github.com/rust-lang/rfcs/pull/3502):
+````rust
+#!/usr/bin/env cargo
+```cargo
+[dependencies]
+clap = { version = "4.2", features = ["derive"] }
+```
+
+use clap::Parser;
+
+#[derive(Parser, Debug)]
+#[clap(version)]
+struct Args {
+    #[clap(short, long, help = "Path to config")]
+    config: Option<std::path::PathBuf>,
+}
+
+fn main() {
+    let args = Args::parse();
+    println!("{:?}", args);
+}
+````
+
+# Motivation
+[motivation]: #motivation
+
+["cargo script"](https://github.com/rust-lang/rfcs/pull/3502) is in need of a syntax for embedding manifests in source.
+See that RFC for its motivations.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Static site generators use a frontmatter syntax to embed metadata in markdown files:
+```md
+---
+name: My Blog Post
+---
+
+## Stuff happened
+
+Hello world!
+```
+
+We are carrying this concept over to Rust with a twist: using code fences which
+will be familiar to Rust developers when documenting their code:
+````rust
+#!/usr/bin/env cargo
+```cargo
+[dependencies]
+clap = { version = "4.2", features = ["derive"] }
+```
+
+use clap::Parser;
+
+#[derive(Parser, Debug)]
+#[clap(version)]
+struct Args {
+    #[clap(short, long, help = "Path to config")]
+    config: Option<std::path::PathBuf>,
+}
+
+fn main() {
+    let args = Args::parse();
+    println!("{:?}", args);
+}
+````
+
+As we work to better understand how tool authors will want to use frontmatter, we are restricting it to just the `cargo` infostring.
+This means users will only be exposed to this within the concept of ["cargo script"](https://github.com/rust-lang/rfcs/pull/3502).
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+When parsing Rust code, after stripping the shebang (`#!`), rustc will strip a code fence:
+- Must be immediately at the top (after shebang stripping), meaning no blank lines
+- Opens with 3+ backticks and "cargo" followed by a newline
+- All content is ignored until a matching number of backticks is found at the start of a line.  It is an error to have anything come between the backticks and the newline.
+
+As cargo will be the first step in the process to parse this,
+the responsibility for high quality error messages will largely fall on cargo.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- A new concept for Rust syntax, adding to overall cognitive load
+- Requires people escape markdown code fences with an extra backtick which they are likely not used to doing (or aware even exists)
```

Comment:
Does this refer to markdown inside multiline TOML strings? It's not entirely clear. Do any of Cargo's manifest fields even support markdown syntax?

Reply:
It may be referring to general usage outside of Rust source files, for example if I wanted to express this syntax here in this comment I would have to escape the backticks somehow (or indent by four spaces, but that doesn't allow syntax highlighting).

Reply:
Ooohh I see. In that case there's also the option of using the (older, I think?) syntax of prefixing the `.rs` snippet with four spaces instead of ````` ```` `````, though I'm not sure it it's as well-known as code fences, and it has the downside of not supporting language tags.

Reply:
394387b5 adds some context.  This is about sharing snippets over github or zulip.  Since you are putting a code fence inside of a code fence, the outer one needs to use 4 backticks